### DISCUSSION TITLE
I am confused (so this is either a bug report or a presentation of my stupidity i guess)

### DISCUSSION THREAD

**Discussion by Kukuschi** [edwardkort/WWIDesigner#134] [Whistle Study] [answered]
I tried the software for a few days now and i am confused. On the one hand it is a great idea but it seems not to work as expected (or at least as i expect.... maybe i am wrong) most of the time.

My goal is to build a Low D whistle. (I am a bad whistle player, mediocre coder and a good (but poor) 3d designer and do-it-yourselfer).
As starting point i used the dimensions from an online whistle calculator and changed one whistle file that is delivered with the software accordingly.
As tuning i used frequencies and fingerings for just the first two octaves (without any cross fingering) to keep it simple.
Looking at the tuning table  (deviation of expected frequency) the whistle from the online calculator was quite a bit off, but not horrible.

I then set all the constraints fitting for a Low D and ran different optimisations (mainly "size + spacing" and "size + spacing global") and looked at the tuning table to see how much it got optimized each time.

To my surprise it will not always get better every time but (very) often the result will be (much) worse then the starting point.
If i was stuck i sligthly changed some values by hand and run again to get new/better results (with some luck).

Optimisation that makes things much worse seem to be quite counterintuitive.
The software also told me, that it found the optimal result, but after i changed one value (e.g. the position of a hole by a few mm within constraints bounds) by hand and ran the optimisation again, the results where sometimes (much) better than this optimum.

That way i was able to get an instrument that is (per tuning table) nowhere more off then 4 cents. (error rating araoun 63 ... whatever that does mean).

I then made a new tuning.
For that i did load the basic one and extended  it by one octave and some cross fingering (but did not change what was already there).
My plan was to keep one best (least deviation) fingering for each note and then further optimize to get a good enough result for the whole whistle.

But... and here i am even more confused: I did not change the initial tuning, just added a few other fingerings and still the tuning table changed quite significantly (all the expected frequencies/deviations were quite a bit worse all of a sudden while - as far as i understand it - they should not have changed at all).

I really love the Idea of the software (and can live with the confusing UI and the strangely laborious workflow) ... but it kinda seems quite a bit buggy and/or unreliable and i am not shure how far it can be trusted. (And that is a shame cause there is nothing similar out there)

---

Edit (attached screenshot as clarification):

![deviation](https://user-images.githubusercontent.com/64613417/134391774-86444192-2d9a-4760-a62c-1a4f69d9e94b.png)

Here is twice the same instrument.
From D4 to C#6 the tuning is identical in both (frequencies and fingering etc.), the only difference is the added fingerings after C#6 in the second table. So afaik the values in the red boxes should be the same for both.

**Comment by bhp1** [comment, marked as answer]
When you added new fingerings, in particular because you added a whole new octave on the top end, WIDesigner changed its assumption of how you wanted to blow the instrument.  To predict the [nominal playing frequency](https://github.com/edwardkort/WWIDesigner/wiki/Working-With-the-Whistle-Study-Model#measuring-frequencies), the WIDesigner whistle model assumes that you will steadily increase the air velocity in the windway as you go from the lowest note on the instrument to the highest.  To reach C#7, you are going to have to blow a lot harder, and the model assumes you want to distribute that across the whole range, rather than a sudden increase around D6 or E6.  This breath curve means you are going to be blowing the notes in the first two octaves harder than you were when you only needed to get up to C#6, so the notes in the first two octaves are going to come out sharper.

You can see this more clearly on [the tuning graph](https://github.com/edwardkort/WWIDesigner/wiki/Evaluating-Instruments-with-the-Tuning-Table-and-Tuning-Graph#the-tuning-graph).  Display the tuning graph for the two-octave tuning file, then for the three-octave tuning file.  Look at the line of small blue circles - that's the breath curve that WIDesigner assumes you are aiming for.  The breath curve for the three-octave tuning is higher than for the two-octave tuning because it has to be to get to C#7, so the predicted tuning is sharper.  The green dots for the first two octaves (or yellow or red markers), which represent the target tuning, will be in the same places on the two graphs.

**Reply by Kukuschi** [reply]
Ah, that makes sense, thanks for the explanation. :)

**Comment by bhp1** [comment]
A word is in order about the optimizers ... The [global optimizers](https://github.com/edwardkort/WWIDesigner/wiki/WIDesigner-Optimizers#optimization-algorithms) attempt an extensive search in the whole search space between the constraints.  They don't use the initial conditions.  If the constraints are broad, they are likely to land in a local optimum that is not as good as the initial conditions.  I suggest you use them only when you _don't_ have a good set of initial conditions, but you do have constraints set fairly tightly around the expected optimum.  The normal optimizers work better than the global optimizers if you do have a reasonably good set of initial conditions.  They will occasionally land on a geometry that is worse than the initial conditions, but not often.

**Reply by Kukuschi** [reply]
Ah, "local optimum" explains a lot i think.
Would an option to make an optimizer explore/search other local optima (showing the so far best result after each run) until the user stops it possible? It could run over night.

**Reply by bhp1** [reply]
The global optimizer tests a broad range of points within the search space (between the constraints), and is unlikely to get caught in a local optimum.  It gives up only when its search strategy shows no signs of improvement over many iterations.  As yet, the programming doesn't offer us a convenient mechanism to interrupt an optimization and keep the best result so far.  If we had such a mechanism, it would be tempting to let the optimizer run all night searching for better solutions.

With the standard optimizer (BOBYQA), you can look for other local optima by changing the starting geometry.  If necessary, you can change the constraints to force it to stay away from a particular local optimum.

To explore other solutions with the global optimizer, change the constraints.  I find it helps to narrow the constraints around where good solutions are likely to be found.
